src/operator/join_range.rs

```diff
@@ -0,0 +1,303 @@
+//! Range-join operators.
+//!
+//! Range-join is a form of non-equi join where each key in the left operand
+//! matches a contiguous range of keys in the right operand.
+//!
+//! Consider two indexed Z-sets `z1` and `z2`, a function `range_func` that
+//! maps a key in `z1` to a half-closed interval of keys `[lower, upper)` in
+//! `z2` and another function `join_func` that, given `(k1, v1)` in `z1` and
+//! `(k2, v2)` in `z2`, returns an iterable collection `c` of output values.
+//! The range-join operator works as follows:
```

Comment:
this is a strange definition. are you sure it's bilinear?
also, how is this used to model more common types of joins like JOIN on T1.x < T2.y?

Reply:
Why do you think it's strange? I'm pretty sure it is bilinear.

> also, how is this used to model more common types of joins like JOIN on T1.x < T2.y?

It's not meant to be the most general equijoin. It can probably be generalized or we may add another operator for that.

Reply:
It is strange because the second function returns a collection. The size of the colection is not related to the size of the inputs

Reply:
Usually the collection is `Option<>`, but it can also be, e.g., `Vec<>`. I don't think this breaks bi-linearity, since each element in the collection will be added with the weight that is the product of input weights.